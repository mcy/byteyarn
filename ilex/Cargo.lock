# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "annotate-snippets"
version = "0.9.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ccaf7e9dfbb6ab22c82e473cd1a8a7bd313c19a5b7e40970f3d89ef5a5c9e81e"
dependencies = [
 "unicode-width",
 "yansi-term",
]

[[package]]
name = "bitflags"
version = "2.4.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "327762f6e5a765692301e5bb513e0d9fef63be86bbc14528052b1cd3e6f03e07"

[[package]]
name = "byteyarn"
version = "0.2.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a7534301c0ea17abb4db06d75efc7b4b0fa360fce8e175a4330d721c71c942ff"

[[package]]
name = "camino"
version = "1.1.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c59e92b5a388f549b863a7bea62612c09f24c8393560709a54558a9abdfb3b9c"

[[package]]
name = "diff"
version = "0.1.13"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "56254986775e3233ffa9c4d7d3faaf6d36a2c09d30b20687e9f88bc8bafc16c8"

[[package]]
name = "ilex"
version = "0.1.0"
dependencies = [
 "annotate-snippets",
 "bitflags",
 "byteyarn",
 "camino",
 "pretty_assertions",
 "unicode-segmentation",
 "unicode-xid",
]

[[package]]
name = "pretty_assertions"
version = "1.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "af7cee1a6c8a5b9208b3cb1061f10c0cb689087b3d8ce85fb9d2dd7a29b6ba66"
dependencies = [
 "diff",
 "yansi",
]

[[package]]
name = "unicode-segmentation"
version = "1.10.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1dd624098567895118886609431a7c3b8f516e41d30e0643f03d94592a147e36"

[[package]]
name = "unicode-width"
version = "0.1.11"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e51733f11c9c4f72aa0c160008246859e340b00807569a0da0e7a1079b27ba85"

[[package]]
name = "unicode-xid"
version = "0.2.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f962df74c8c05a667b5ee8bcf162993134c104e96440b663c8daa176dc772d8c"

[[package]]
name = "winapi"
version = "0.3.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5c839a674fcd7a98952e593242ea400abe93992746761e38641405d28b00f419"
dependencies = [
 "winapi-i686-pc-windows-gnu",
 "winapi-x86_64-pc-windows-gnu",
]

[[package]]
name = "winapi-i686-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ac3b87c63620426dd9b991e5ce0329eff545bccbbb34f3be09ff6fb6ab51b7b6"

[[package]]
name = "winapi-x86_64-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "712e227841d057c1ee1cd2fb22fa7e5a5461ae8e48fa2ca79ec42cfc1931183f"

[[package]]
name = "yansi"
version = "0.5.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "09041cd90cf85f7f8b2df60c646f853b7f535ce68f85244eb6731cf89fa498ec"

[[package]]
name = "yansi-term"
version = "0.1.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "fe5c30ade05e61656247b2e334a031dfd0cc466fadef865bdcdea8d537951bf1"
dependencies = [
 "winapi",
]
