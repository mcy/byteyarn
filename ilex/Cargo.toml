[package]
name = "ilex"
version = "0.4.1"
description = "quick and easy lexers for C-like languages"

edition.workspace = true
authors.workspace = true
homepage.workspace = true
repository.workspace = true
keywords.workspace = true
license.workspace = true

[dependencies]
byteyarn = { version = "0.4", path = "../byteyarn" }
twie = { version = "0.4", path = "../twie" }

annotate-snippets = "0.10.0"
camino = "1.1.6"
dashmap = "5.5.3"
num-traits = "0.2.17"
similar-asserts = "1.5.0"
rustc_apfloat = "0.2.0"  # By eddyb's recommendation.
unicode-xid = "0.2.4"
