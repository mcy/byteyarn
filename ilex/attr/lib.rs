//! Implementation detail of `ilex`.

use proc_macro::TokenStream;

// This helper exists only to make the #[spec] field attributes inert.
#[doc(hidden)]
#[proc_macro_derive(derive_hack, attributes(named, rule))]
pub fn derive(_: TokenStream) -> TokenStream {
  TokenStream::new()
}

proc2decl::bridge! {
  /// Generates a lexer spec struct.
  ///
  /// This macro generates the type of struct described in the
  /// [crate documentation][crate]. The syntax is as follows.
  ///
  /// ```ignore
  /// use ilex::rule::Keyword;
  /// use ilex::Lexeme;
  ///
  /// /// My cool spec.
  /// #[ilex::spec]
  /// struct MySpec {
  ///   #[named("...")]
  ///   #[rule(/* ... */)]
  ///   dollar: Lexeme<Keyword> = "$",
  /// }
  /// ```
  ///
  /// The type of each field must be a [`Lexeme`] with a [`Rule`] type as its
  /// parameter. There are two special attributes that can follow.
  ///
  /// - `#[named]` makes the rule into a *named* rule. This name can be used by
  ///   diagnostics, and corresponds to calling `Spec::named_rule()`.
  ///
  /// - `#[rule]` is the value to use to construct the rule, which must be
  ///   `Into<R>`, where `R` is the type inside `Lexeme` (so, above, the rule
  ///   value must be `Into<Keyword>`). By default, this value is the name of the
  ///   rule, to make the common case of declaring a keyword as simple as writing
  ///   `nullptr: Lexeme<Keyword>`, assuming Rust itself doesn't already use that
  ///   keyword.
  ///
  /// Note that *order matters* for the fields: when breaking a tie between two
  /// potential tokens of the same length, the first one in the struct will win.
  /// In practice, this means you should put keywords before identifiers.
  ///
  /// Additionally, the following functions will be defined for the `MySpec` type.
  ///
  /// ```
  /// # struct Spec;
  /// # struct MySpec;
  /// # fn norun(_: i32) {
  /// impl MySpec {
  ///   /// Gets the global instance of this spec.
  ///   pub fn get() -> &'static Self {
  ///     // ...
  /// #   todo!()
  ///   }
  ///
  ///   /// Gets the actual compiled spec.
  ///   pub fn spec(&self) -> &Spec {
  ///     // ...
  /// #   todo!()
  ///   }
  /// }
  /// # }
  /// ```
  ///
  // God cross-trait links suck.
  /// [`Lexeme`]: https://docs.rs/ilex/latest/ilex/struct.Lexeme.html
  /// [`Rule`]: https://docs.rs/ilex/latest/ilex/rule/trait.Rule.html
  /// [crate]: https://docs.rs/ilex
  macro #[spec] => ilex::__spec__;
}
