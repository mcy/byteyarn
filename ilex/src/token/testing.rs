//! Lexer testing helpers.
//!
//! This type provides testing-oriented matchers for matching on a
//! [`TokenStream`][`crate::token::Stream`].
//!
//! These matchers are intended for writing *tests*. To write a parser, you\
//! should use [`Cursor`][crate::token::Cursor] instead.

use std::fmt;
use std::fmt::DebugStruct;
use std::fmt::Display;
use std::ops::Range;

use std::format_args as f;

use byteyarn::Yarn;

use crate::file::Context;
use crate::file::Span;
use crate::file::Spanned;
use crate::lexer::rule;
use crate::lexer::Lexeme;
use crate::token;
use crate::token::Any;
use crate::token::Cursor;
use crate::token::Sign;

/// Matches a token cursor against a list of token matchers.
///
/// Returns `Ok(())` when they match, otherwise returns a pretty-printed error
/// showing what went wrong.
pub fn recognize_tokens<'a>(
  ctx: &Context,
  stream: Cursor,
  matchers: impl IntoIterator<Item = &'a Matcher>,
) -> Result<(), impl fmt::Debug> {
  let mut state = MatchState {
    ctx,
    errors: String::new(),
    stack: Vec::new(),
    error_count: 0,
  };

  struct DebugBy(String);
  impl fmt::Debug for DebugBy {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
      f.write_str(&self.0)
    }
  }

  zip_eq(&mut state, matchers, stream, |state, ours, theirs| {
    ours.recognizes(state, theirs, ctx)
  });
  state.finish().map_err(DebugBy)
}

/// A matcher for a chunk of text from the input source.
///
/// This is slightly more general than a span, since it can specify the content
/// of the text and the offsets separately, and optionally. `Text` values are
/// intended to *recognize* various spans.
///
/// `&str` and `Range<usize>` are both convertible to `Text`.
pub struct Text {
  text: Option<Yarn>,
  range: Option<Range<usize>>,
}

impl Text {
  /// Returns a matcher that recognizes all spans.
  pub fn any() -> Self {
    Text {
      text: None,
      range: None,
    }
  }

  /// Returns a matcher that recognizes spans with the given text.
  pub fn new(text: impl Into<Yarn>) -> Self {
    Text {
      text: Some(text.into()),
      range: None,
    }
  }

  /// Returns a matcher that recognizes spans with the given byte range.
  pub fn range(range: Range<usize>) -> Self {
    Text {
      text: None,
      range: Some(range),
    }
  }

  /// Returns a matcher that recognizes spans with the given byte range and
  /// text.
  pub fn text_and_range(text: impl Into<Yarn>, range: Range<usize>) -> Self {
    Text {
      text: Some(text.into()),
      range: Some(range),
    }
  }

  /// Returns whether this span recognizes a particular span.
  fn recognizes(&self, span: Span, ctx: &Context) -> bool {
    !self
      .text
      .as_ref()
      .is_some_and(|text| text != span.text(ctx))
      && !self
        .range
        .as_ref()
        .is_some_and(|range| range != &span.range(ctx).unwrap())
  }
}

impl<Y: Into<Yarn>> From<Y> for Text {
  fn from(value: Y) -> Self {
    Text::new(value.into())
  }
}

impl fmt::Debug for Text {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    match (&self.text, &self.range) {
      (Some(text), Some(range)) => write!(f, "{text:?} @ {range:?}"),
      (Some(text), None) => fmt::Debug::fmt(text, f),
      (None, Some(range)) => write!(f, "<any> @ {range:?}"),
      (None, None) => f.write_str("<any>"),
    }
  }
}

pub struct Matcher {
  which: Option<Lexeme<rule::Any>>,
  span: Text,
  comments: Vec<Text>,
  kind: Kind,
}

enum Kind {
  Eof,
  Unexpected,
  Keyword,
  Ident {
    name: Text,
    prefix: Option<Text>,
    suffix: Option<Text>,
  },
  Quoted {
    content: Vec<token::Content<Text>>,
    delims: (Text, Text),
    prefix: Option<Text>,
    suffix: Option<Text>,
  },
  Digital {
    mant: DigitalMatcher,
    exps: Vec<DigitalMatcher>,
    suffix: Option<Text>,
  },
  Delimited {
    delims: (Text, Text),
    tokens: Vec<Matcher>,
  },
}

#[derive(Debug)]
struct DigitalMatcher {
  radix: u8,
  sign: Option<(Sign, Text)>,
  digits: Vec<Text>,
  prefix: Option<Text>,
}

impl Lexeme<rule::Keyword> {
  pub fn matcher(self, text: impl Into<Text>) -> Matcher {
    Matcher {
      which: Some(self.any()),
      span: text.into(),
      comments: Vec::new(),
      kind: Kind::Keyword,
    }
  }
}

impl Lexeme<rule::Ident> {
  pub fn matcher(self, name: impl Into<Text>) -> Matcher {
    Matcher {
      which: Some(self.any()),
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Ident {
        name: name.into(),
        prefix: None,
        suffix: None,
      },
    }
  }
}

impl<Y: Into<Yarn>> From<Y> for token::Content<Text> {
  fn from(value: Y) -> Self {
    Self::Lit(value.into().into())
  }
}

impl Lexeme<rule::Quoted> {
  pub fn matcher<C: Into<token::Content<Text>>>(
    self,
    quotes: (impl Into<Text>, impl Into<Text>),
    content: impl IntoIterator<Item = C>,
  ) -> Matcher {
    Matcher {
      which: Some(self.any()),
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Quoted {
        content: content.into_iter().map(C::into).collect(),
        delims: (quotes.0.into(), quotes.1.into()),
        prefix: None,
        suffix: None,
      },
    }
  }
}

impl Lexeme<rule::Digital> {
  pub fn matcher<B, T>(self, radix: u8, blocks: B) -> Matcher
  where
    B: IntoIterator<Item = T>,
    T: Into<Text>,
  {
    Matcher {
      which: Some(self.any()),
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Digital {
        mant: DigitalMatcher {
          radix,
          sign: None,
          digits: blocks.into_iter().map(Into::into).collect(),
          prefix: None,
        },
        exps: Vec::new(),
        suffix: None,
      },
    }
  }
}

impl Lexeme<rule::Bracket> {
  pub fn matcher(
    self,
    brackets: (impl Into<Text>, impl Into<Text>),
    contents: Vec<Matcher>,
  ) -> Matcher {
    Matcher {
      which: Some(self.any()),
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Delimited {
        tokens: contents,
        delims: (brackets.0.into(), brackets.1.into()),
      },
    }
  }
}

impl Matcher {
  pub fn eof() -> Self {
    Matcher {
      which: Some(Lexeme::eof().any()),
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Eof,
    }
  }

  pub fn unexpected() -> Self {
    Matcher {
      which: None,
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Unexpected,
    }
  }

  pub fn span(mut self, text: impl Into<Text>) -> Self {
    self.span = text.into();
    self
  }

  pub fn prefix(mut self, text: impl Into<Text>) -> Self {
    match &mut self.kind {
      Kind::Ident { prefix, .. } | Kind::Quoted { prefix, .. } => {
        *prefix = Some(text.into());
      }
      Kind::Digital { mant, .. } => mant.prefix = Some(text.into()),
      _ => unreachable!(),
    }

    self
  }

  pub fn suffix(mut self, text: impl Into<Text>) -> Self {
    match &mut self.kind {
      Kind::Ident { suffix, .. }
      | Kind::Quoted { suffix, .. }
      | Kind::Digital { suffix, .. } => {
        *suffix = Some(text.into());
      }
      _ => unreachable!(),
    }

    self
  }

  pub fn sign(mut self, value: Sign, text: impl Into<Text>) -> Self {
    match &mut self.kind {
      Kind::Digital { mant, .. } => mant.sign = Some((value, text.into())),
      _ => unreachable!(),
    }

    self
  }

  pub fn exponent<T: Into<Text>>(
    mut self,
    radix: u8,
    delim: impl Into<Text>,
    sign: Option<(Sign, Text)>,
    blocks: impl IntoIterator<Item = T>,
  ) -> Self {
    match &mut self.kind {
      Kind::Digital { exps, .. } => exps.push(DigitalMatcher {
        radix,
        sign,
        digits: blocks.into_iter().map(Into::into).collect(),
        prefix: Some(delim.into()),
      }),
      _ => unreachable!(),
    }

    self
  }

  pub fn comments<I>(mut self, iter: I) -> Self
  where
    I: IntoIterator,
    I::Item: Into<Text>,
  {
    self.comments.extend(iter.into_iter().map(Into::into));
    self
  }

  fn recognizes(&self, state: &mut MatchState, tok: token::Any, ctx: &Context) {
    state.match_spans("token span", &self.span, tok.span(ctx));

    zip_eq(state, &self.comments, tok.comments(ctx), |state, t, s| {
      state.match_spans("comment", t, s);
    });

    match (&self.kind, tok) {
      (Kind::Eof, Any::Eof(..))
      | (Kind::Unexpected, Any::Unexpected(..))
      | (Kind::Keyword, Any::Keyword(..)) => {}
      (
        Kind::Ident {
          name,
          prefix,
          suffix,
        },
        Any::Ident(tok),
      ) => {
        state.match_spans("identifier name", name, tok.name());
        state.match_options("prefix", prefix.as_ref(), tok.prefix());
        state.match_options("suffix", suffix.as_ref(), tok.suffix());
      }
      (
        Kind::Quoted {
          delims,
          content,
          prefix,
          suffix,
        },
        Any::Quoted(tok),
      ) => {
        let (open, close) = tok.delimiters();
        state.match_spans("open quote", &delims.0, open);
        state.match_spans("close quote", &delims.1, close);
        state.match_options("prefix", prefix.as_ref(), tok.prefix());
        state.match_options("suffix", suffix.as_ref(), tok.suffix());

        zip_eq(
          state,
          content,
          tok.raw_content(),
          |state, ours, theirs| match (ours, theirs) {
            (token::Content::Lit(t), token::Content::Lit(s)) => {
              state.match_spans("string content", t, s)
            }
            (token::Content::Esc(t, ours), token::Content::Esc(s, theirs)) => {
              state.match_spans("string escape", t, s);
              if ours != &theirs {
                state.error(f!(
                  "wrong decoded escape; want {:?}, got {:?}",
                  ours,
                  theirs
                ));
              }
            }
            _ => state.error("mismatched string content types"),
          },
        );
      }
      (Kind::Digital { mant, exps, suffix }, Any::Digital(tok)) => {
        let recognize = |state: &mut MatchState,
                         mch: &DigitalMatcher,
                         tok: token::Digital| {
          if mch.radix != tok.radix() {
            state.error(f!(
              "wrong radix; want {:?}, got {:?}",
              mch.radix,
              tok.radix()
            ));
          }
          state.match_any_options(
            "sign",
            mch.sign.as_ref().map(|(s, _)| s),
            tok.sign(),
            |&a, b| a == b,
          );
          state.match_options(
            "sign span",
            mch.sign.as_ref().map(|(_, sp)| sp),
            tok.sign_span(),
          );
          state.match_options("prefix", mch.prefix.as_ref(), tok.prefix());
          zip_eq(state, &mch.digits, tok.digit_blocks(), |state, t, s| {
            state.match_spans("digit block", t, s);
          });
        };

        recognize(state, mant, tok);
        zip_eq(state, exps, tok.exponents(), |state, t, s| {
          recognize(state, t, s);
        });

        state.match_options("suffix", suffix.as_ref(), tok.suffix());
      }
      (Kind::Delimited { delims, tokens }, Any::Bracket(tok)) => {
        state.match_spans("open delimiter", &delims.0, tok.open());
        state.match_spans("close delimiter", &delims.1, tok.close());

        zip_eq(state, tokens, tok.contents(), |state, ours, theirs| {
          ours.recognizes(state, theirs, ctx)
        });
      }
      _ => state.error("mismatched token types"),
    }
  }
}

impl fmt::Debug for Matcher {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    let print_spans = matches!(
      std::env::var("ILEX_SPANS").as_deref(),
      Ok("ranges" | "text")
    );

    let req_field = |d: &mut DebugStruct, name, span| {
      if print_spans {
        d.field(name, span);
      }
    };

    let opt_field = |d: &mut DebugStruct, name, span: &Option<Text>| {
      if print_spans && span.is_some() {
        d.field(name, span.as_ref().unwrap());
      }
    };

    let vec_field = |d: &mut DebugStruct, name, spans: &Vec<Text>| {
      if !spans.is_empty() {
        d.field(name, spans);
      }
    };

    let name = match &self.kind {
      Kind::Eof => "Eof",
      Kind::Unexpected => "Unexpected",
      Kind::Keyword => "Keyword",
      Kind::Ident { .. } => "Ident",
      Kind::Quoted { .. } => "Quoted",
      Kind::Digital { .. } => "Digital",
      Kind::Delimited { .. } => "Delimited",
    };

    let name = match self.which {
      Some(l) => format!("{name}({})", l.index()),
      None => name.into(),
    };

    let mut d = f.debug_struct(&name);

    match &self.kind {
      Kind::Ident {
        name,
        prefix,
        suffix,
      } => {
        req_field(&mut d, "name", name);
        opt_field(&mut d, "prefix", prefix);
        opt_field(&mut d, "suffix", suffix);
      }
      Kind::Quoted {
        content,
        delims,
        prefix,
        suffix,
      } => {
        d.field("content", content);
        req_field(&mut d, "open", &delims.0);
        req_field(&mut d, "close", &delims.1);
        opt_field(&mut d, "prefix", prefix);
        opt_field(&mut d, "suffix", suffix);
      }
      Kind::Digital { mant, exps, suffix } => {
        d.field("mant", mant).field("exps", exps);
        opt_field(&mut d, "suffix", suffix);
      }
      Kind::Delimited { delims, tokens } => {
        req_field(&mut d, "open", &delims.0);
        req_field(&mut d, "close", &delims.1);
        d.field("tokens", tokens);
      }
      _ => {}
    };

    req_field(&mut d, "span", &self.span);
    vec_field(&mut d, "comments", &self.comments);
    d.finish()
  }
}

struct MatchState<'a> {
  ctx: &'a Context,
  errors: String,
  stack: Vec<usize>,
  error_count: usize,
}

impl MatchState<'_> {
  fn error(&mut self, msg: impl Display) {
    use std::fmt::Write;

    self.error_count += 1;
    if self.error_count > 10 {
      return;
    }

    self.errors.push_str("at stream");
    for i in &self.stack {
      let _ = write!(self.errors, "[{}]", i);
    }
    let _ = writeln!(self.errors, ": {msg}");
  }

  fn match_spans(&mut self, what: &str, text: &Text, span: Span) {
    if !text.recognizes(span, self.ctx) {
      self.error(f!("wrong {what}; want {:?}, got {:?}", text, span));
    }
  }

  fn match_options(
    &mut self,
    what: &str,
    text: Option<&Text>,
    span: Option<Span>,
  ) {
    if text.is_none() && span.is_none() {
      return;
    }

    if !text
      .zip(span)
      .is_some_and(|(t, s)| t.recognizes(s, self.ctx))
    {
      self.error(f!("wrong {what}; want {:?}, got {:?}", text, span));
    }
  }

  fn match_any_options<T: fmt::Debug, U: fmt::Debug>(
    &mut self,
    what: &str,
    text: Option<T>,
    span: Option<U>,
    eq: impl FnOnce(&T, &U) -> bool,
  ) {
    if text.is_none() && span.is_none() {
      return;
    }

    if !text
      .as_ref()
      .zip(span.as_ref())
      .is_some_and(|(t, s)| eq(t, s))
    {
      self.error(f!("wrong {what}; want {:?}, got {:?}", text, span));
    }
  }

  fn finish(mut self) -> Result<(), String> {
    use std::fmt::Write;

    if self.error_count > 10 {
      let _ =
        writeln!(self.errors, "... and {} more errors", self.error_count - 1);
    }

    if self.error_count > 0 {
      return Err(self.errors);
    }
    Ok(())
  }
}

fn zip_eq<Ours: IntoIterator, Theirs: IntoIterator>(
  state: &mut MatchState,
  ours: Ours,
  theirs: Theirs,
  mut cb: impl FnMut(&mut MatchState, Ours::Item, Theirs::Item),
) {
  let mut ours = ours.into_iter();
  let mut theirs = theirs.into_iter();
  state.stack.push(0);
  loop {
    let ours = ours.next();
    let theirs = theirs.next();
    if ours.is_none() && theirs.is_none() {
      state.stack.pop();
      break;
    }

    if let (Some(ours), Some(theirs)) = (ours, theirs) {
      cb(state, ours, theirs);

      *state.stack.last_mut().unwrap() += 1;
      continue;
    }

    state.error("unequal lengths");
    state.stack.pop();
    break;
  }
}
