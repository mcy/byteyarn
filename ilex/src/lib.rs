//! Quick and easy lexing for C-like languages.
//!
//! This crate provides a highly general lexer for a C-like language, as well as
//! comprehensive span support. This library is based off of a specific kind of
//! parser I have re-written, verbatim, many times over in my career.
//!
//! The goals of this library are:
//!
//! - Predictably greedy behavior. Always parse the longest token at any
//!   particular position.
//!
//! - Easy to set up. Writing lexers is a bunch of pain, and they all look the
//!   same more-or-less, and you want to be "in and out".
//!
//! - It can lex a reasonably large number of grammars. It should be able to do
//!   any language with a cursory resemblance to C, such as Rust, JavaScript,
//!   LLVM IR, Go, Protobuf, and so on.
//!
//! - Unicode XID support (this, in particular, means this lexer doesn't use
//!   action tables or finite automata).
//!
//! - Diagnostics and span management. The lexer should be able to generate
//!   diagnostics, and this API is exposed for tools built on top of the lexer
//!   to emit diagnostics.
//!
//! # Quick Start
//!
//! This library is *not* a lexer generator, in that it does not generate code.
//! Instead, you define a lexer specification, which is compiled into a
//! structure that is kind of, sort of like can a lexer action table, but
//! somewhat more optimized for parsing brackets, numbers, strings, identifiers,
//! keywords, and so on.
//!
//! For example, he's what a lexer for JSON looks like.
//!
/*! ```
use ilex::Lexeme;
use ilex::rule::Keyword;
use ilex::rule::Bracket;
use ilex::rule::Quoted;
use ilex::rule::Digital;
use ilex::rule::Digits;

// This is a spec builder. You give it rule definitions, and it produces
// "lexemes", which are IDs for later recalling which rule a token was matched
// by.
let mut spec = ilex::Spec::builder();

// It is convenient to wrap all the lexemes in a struct, so that they can be
// recalled by name elsewhere in the parsing stack.
struct Json {
  comma: Lexeme<Keyword>,
  colon: Lexeme<Keyword>,
  minus: Lexeme<Keyword>,
  true_: Lexeme<Keyword>,
  false_: Lexeme<Keyword>,
  null: Lexeme<Keyword>,

  array: Lexeme<Bracket>,
  object: Lexeme<Bracket>,
  string: Lexeme<Quoted>,
  number: Lexeme<Digital>,

  spec: ilex::Spec,
}

let json = Json {
  // Keywords are any string, so "punctuation" also a keyword.
  comma: spec.rule(Keyword::new(",")),
  colon: spec.rule(Keyword::new(".")),
  minus: spec.rule(Keyword::new("-")),
  true_: spec.rule(Keyword::new("true")),
  false_: spec.rule(Keyword::new("false")),
  null: spec.rule(Keyword::new("null")),

  // "Bracket" are special kinds of matched rules that must appear in
  // opposition. I.e., matched brackets. Brackets can actually be
  // non-context-free, since there is explicit support for Rust-style AND
  // C++-style strings, neither of which are context free.
  array: spec.named_rule("array", Bracket::from(("[", "]"))),
  object: spec.named_rule("object", Bracket::from(("{", "}"))),

  // A "quoted" or "quoted rule" is a generalization of a string. It is
  // matched delimiters and a collection of escapes that could appear within it.
  // Escapes can be reasonably complicated and can have their own parsing
  // functions.
  string: spec.named_rule(
    "string",
    Quoted::new('"')
      .invalid_escape(r"\")
      .escapes([
        "\\\"", r"\\", r"\/",
        r"\b", r"\f",  r"\n", r"\t", r"\r",
      ])
      .fixed_length_escape(r"\u", 4),
  ),

  // A digital rule is something that, resembles a number! Digitals are of a
  // specific radix (so you'll need separate rules for your 123 and 0xbeef), and
  // can have arbitrary numbers of decimal points: you could lex something like
  // 1.0.0 if you wanted. They can also "exponents", for lexing floats.
  number: spec.rule(
    Digital::new(10)
      .minus()
      .point_limit(0..2)
      .exponents(["e", "E"], Digits::new(10).plus().minus()),
  ),

  // Wrap it up and compile the spec.
  spec: spec.compile(),
};
```*/
//! This is the intended idiom for using `ilex`; there is a convenient macro
//! for doing this in a single step.
//!
/*! ```
use ilex::rule::Keyword;
use ilex::rule::Bracket;
use ilex::rule::Quoted;
use ilex::rule::Digital;
use ilex::rule::Digits;

ilex::spec! {
  struct Json {
    comma: Keyword = ',',
    colon: Keyword = ':',
    minus: Keyword = '-',
    true_: Keyword = "true",
    false_: Keyword = "false",
    null: Keyword = "null",

    #[named] array: Bracket = ('[', ']'),
    #[named] object: Bracket = ('{','}'),
    #[named] string: Quoted = Quoted::new('"')
      .invalid_escape(r"\")
      .escapes([
        "\\\"", r"\\", r"\/",
        r"\b", r"\f",  r"\n", r"\t", r"\r",
      ])
      .fixed_length_escape(r"\u", 4),

    #[named] number: Digital = Digital::new(10)
      .minus()
      .point_limit(0..2)
      .exponents(["e", "E"], Digits::new(10).plus().minus()),
  }
}

let json = Json::get();
let spec = json.spec();

let my_lexeme = json.object;  // Etc.
```*/
//!
//! Other examples of specs can be found in the `tests/` directory. Once you
//! have a spec, you can start lexing some files. For this you'll need a file
//! context. The [`Context`] tracks all files that are being parsed as a part
//! of a parsing session, and is used for looking up the contents of spans.
//!
//! ```
//! use ilex::report;
//! # fn ignore(_: i32) {
//!
//! ilex::spec! {
//!   struct Json {
//!     // As above...
//!   }
//! }
//!
//! // Set up a source context. This tracks all of the source files
//! // we're working with (so source spans can be tiny indices).
//! let mut ctx = ilex::Context::new();
//! let report = ctx.new_report();
//!
//! // Read a file from disk, and lex it with the `json` spec from above.
//! let file = ctx.open_file("my_cool_file.json", &report).unwrap();
//! let tokens = file.lex(Json::get().spec(), &report);
//! # }
//! ```
//!
//! `tokens` here is a [`token::Stream`], which is a tree, since some tokens
//! (delimiters key among them) can contain more tokens *within* them. This is
//! as far as `ilex` will take you.
//!
//! [`ice::handle()`] helps set up this boilerplate and handles generating error
//! messages for ICEs.

#![deny(rustdoc::broken_intra_doc_links)]
#![warn(missing_docs)]

use std::fmt;

macro_rules! bug {
  ($fmt:literal $($arg:tt)*) => {{
    panic!(concat!("ilex: ", $fmt, "; this is a bug") $($arg)*)
  }};
}

pub(crate) use format_args as f;

mod file;
mod rt;
mod spec;

pub mod fp;
pub mod ice;
pub mod report;
pub mod rule;
pub mod testing;
pub mod token;

#[cfg(not(test))]
pub use {
  crate::{
    file::Context,
    file::File,
    file::{Span, SpanId, Spanned},
    report::{Fatal, Report},
    rule::Rule,
    spec::{Lexeme, Spec, SpecBuilder},
    token::Token,
  },
  byteyarn::Yarn,
};

/// The error returned by [`TryFrom`] implementations in this crate.
pub struct WrongKind {
  want: &'static str,
  got: &'static str,
}

impl fmt::Debug for WrongKind {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    write!(f, "wrong kind: want `{}`, got `{}`", self.want, self.got)
  }
}

/// Uninhabited type in lieu of the never type `!`.
///
/// Converts to everything, not constructible.
#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct Never(Void);
#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
enum Void {}

impl Never {
  /// Constructs any value from a `Never`, which cannot be constructed in a
  /// well-formed program.
  #[allow(clippy::wrong_self_convention)]
  fn from_nothing_anything(self) -> ! {
    match self.0 {}
  }
}

fn plural<T: Eq + From<u8>>(count: T) -> &'static str {
  if count == 1.into() {
    ""
  } else {
    "s"
  }
}
