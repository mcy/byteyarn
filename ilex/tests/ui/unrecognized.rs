use ilex::rule::*;
use ilex::testing;
use ilex::Context;

ilex::spec! {
  struct Spec {
    kw: Keyword = "null",
    cm: Bracket = ("[", "]")
  }
}

#[test]
fn unrecognized() {
  let ctx = Context::new();
  let report = ctx.new_report();
  let _ = ctx
    .new_file("<input>", "multiple, null, [unrecognized], chunks!~  ")
    .lex(Spec::get().spec(), &report);

  testing::check_report(&report, "tests/ui/goldens/unrecognized.stdout");
}
