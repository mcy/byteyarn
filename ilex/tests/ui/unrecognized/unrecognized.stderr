error: unrecognized characters
 --> unrecognized.txt:1:1
  |
1 | multiple, null, [unrecognized], chunks!~  
  | ^^^^^^^^^
  |
  = note: reported at: ilex/src/rt/mod.rs:36:8

error: unrecognized character
 --> unrecognized.txt:1:15
  |
1 | multiple, null, [unrecognized], chunks!~  
  |               ^
  |
  = note: reported at: ilex/src/rt/mod.rs:36:8

error: unrecognized characters
 --> unrecognized.txt:1:18
  |
1 | multiple, null, [unrecognized], chunks!~  
  |                  ^^^^^^^^^^^^
  |
  = note: reported at: ilex/src/rt/mod.rs:36:8

error: unrecognized character
 --> unrecognized.txt:1:31
  |
1 | multiple, null, [unrecognized], chunks!~  
  |                               ^
  |
  = note: reported at: ilex/src/rt/mod.rs:36:8

error: unrecognized characters
 --> unrecognized.txt:1:33
  |
1 | multiple, null, [unrecognized], chunks!~  
  |                                 ^^^^^^^^
  |
  = note: reported at: ilex/src/rt/mod.rs:36:8

error: aborting due to 5 errors
